nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,357 @@
+                beg_idx = i
+                end_idx = i + response_size
+                base_weight = -updated_weight[beg_idx:end_idx] - batch["log(Q(y|a,x))"][beg_idx:end_idx].cuda()
+                # for numerical stability, subtract the max value
+                # calculate the probability vector
+                base_weight = base_weight - base_weight.max()
+                base_weight_p = torch.exp(base_weight)
+                # noramliize the weight
+                base_weight_p = base_weight_p / base_weight_p.sum()
+                batch["base_weight"][beg_idx:end_idx] = base_weight_p
+                ws_weight = batch["ws"][beg_idx:end_idx].cuda()
+                batch["weight"][beg_idx:end_idx] = ws_weight - base_weight_p
+                # group size is the number of responses generated
+                loss_mbs = int(self.cfg.data.steerlm2_micro_batch_size)
+
+                num_of_microbatches = gbs // loss_mbs
+                # to compute the loss for each of the group batch,
+                # loss L_i^{jb}  is the loss element for batch $b$, microbatch item $j$ and $i$ item in the microbatch
+                # the loss we need is:  sum_{b=0}^{num_batches} sum_{j=0}^{num_of_microbatches} sum_{i=0}^{mbs} (L_i^{jb}) / num_batches ,
+                # while the fwd-bwd function is used to compute the loss of average microbatch loss
+                # i.e. L = sum_{b=0}^{num_batches} sum_{j=0}^{num_of_microbatches} [sum_{i=0}^{mbs} (L_i^{jb})] / (num_of_microbatches)
+                # to makie it equivalent to the loss we care, we need to scale loss by num_of_microbatches / num_batches
+                # i.e. L = sum_{b=0}^{num_batches} sum_{j=0}^{num_of_microbatches} [(num_of_microbatches)/num_batches * sum_{i=0}^{mbs} (L_i^{jb})] / (num_batches)
+                # also if we want to average the loss weighted by the inverse of number of loss_mass=1 tokens in the batch
+                batch["avg_num_valid_tokens_in_ub"][beg_idx:end_idx] = batch["loss_mask"][beg_idx:end_idx].sum(
+                    axis=-1
+                ).float().mean().repeat(response_size) / (num_of_microbatches / num_of_batches)
+                # compute the kl divergence between group['ws'] and group['base_weight']
+                # for numerical stability, add a small value to the base_weight_p
+                base_weight_p = base_weight_p + 1e-8
+                ws_weight = ws_weight + 1e-8
+                distance = torch.sum(ws_weight * torch.log(ws_weight / base_weight_p))
+                distances.append(distance)
+            # get average distance
+            distance = torch.stack(distances).mean()
+            distance = average_losses_across_data_parallel_group([distance])
+
+        set_sync_funcs(self, forward_only)
+        dp_size = int(parallel_state.get_data_parallel_world_size())
+        gbs = dp_size * gbs
+        mbs = int(self.cfg.data.steerlm2_micro_batch_size)
+        configure_batch_sizes(mbs=mbs, gbs=gbs, dp=dp_size)
```

Comment:
no need to do this if you follow above and don't change globals. Also why not use model.micro_batch_size directly?

Reply:
done